pkg/grpc/api.go

```diff
@@ -1325,20 +1320,66 @@ func (a *api) SubscribeConfigurationAlpha1(request *runtimev1pb.SubscribeConfigu
 		serverStream: configurationServer,
 	}
 
-	ctx := context.TODO()
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
 	// TODO(@laurence) deal with failed subscription and retires
 	err = store.Subscribe(ctx, req, handler.updateEventHandler)
 	if err != nil {
+		err = status.Errorf(codes.InvalidArgument, messages.ErrConfigurationSubscribe, req.Keys, request.StoreName, err.Error())
 		apiServerLogger.Debug(err)
-		a.configurationSubscribeLock.Unlock()
 		return err
 	}
+	stop := make(chan struct{})
+	a.configurationSubscribeLock.Lock()
+	a.configurationSubscribe[getConfigSubscribeUniqueKey(request.GetStoreName(), request.GetKeys())] = stop
+	a.configurationSubscribeLock.Unlock()
+	<-stop
+	return nil
+}
+
+func (a *api) UnSubscribeConfigurationAlpha1(ctx context.Context, request *runtimev1pb.UnSubscribeConfigurationRequest) (*runtimev1pb.UnSubscribeConfigurationResponse, error) {
```

Comment:
If the same client subscribe twice with the same store and the same key set, but the lifecycle of the subscriptions are different deliberately, for example, one subscription will keep alive to the end of the application, while the other subscription (with the same interested set to the same store) probably starts and ends under certain circumstance, then the current UnsubscribeConfigurationAlpha1 is problematic, since it will end both subscriptions which is a unwanted side effect.

If we introduce an id in the return values of SubscribeConfigurationAlpha1, and then use the id to unsubscribe, I believe it would solve the problem mentioned above.

Reply:
Thanks, I mentioned your great idea in [issue](https://github.com/dapr/dapr/issues/3906).
 As gRPC's streaming invocation, the response can't be returned directly, but be returned in async streaming response  , so I think putting the subscribe id in each response is ideal. 
In dapr runtime api level, the id is returned as I explained, as gRPC limitation, but in component level, I put the subscrption id into return value of configuration API interface, which is here: https://github.com/dapr/components-contrib/pull/1483/files
@beiwei30